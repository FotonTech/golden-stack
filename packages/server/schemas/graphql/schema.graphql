"""Represents Event"""
type Event implements Node {
  """The ID of an object"""
  id: ID!

  """MongoDB _id"""
  _id: String!
  title: String
  description: String
  address: String!
  date: String
  createdAt: String
  updatedAt: String
}

input EventAddInput {
  title: String!
  description: String!
  address: String
  date: String
  clientMutationId: String
}

type EventAddPayload {
  eventEdge: EventEdge
  error: String
  clientMutationId: String
}

"""A connection to a list of items."""
type EventConnection {
  """Number of items in this connection"""
  count: Int!

  """
  A count of the total number of objects in this connection, ignoring pagination.
    This allows a client to fetch the first five objects by passing "5" as the
    argument to "first", then fetch the total count so it could display "5 of 83",
    for example.
  """
  totalCount: Int!

  """Offset from start"""
  startCursorOffset: Int!

  """Offset till end"""
  endCursorOffset: Int!

  """Information to aid in pagination."""
  pageInfo: PageInfoExtended!

  """A list of edges."""
  edges: [EventEdge]!
}

"""An edge in a connection."""
type EventEdge {
  """The item at the end of the edge"""
  node: Event!

  """A cursor for use in pagination"""
  cursor: String!
}

type Mutation {
  EventAdd(input: EventAddInput!): EventAddPayload
}

"""An object with an ID"""
interface Node {
  """The id of the object."""
  id: ID!
}

"""Information about pagination in a connection."""
type PageInfoExtended {
  """When paginating forwards, are there more items?"""
  hasNextPage: Boolean!

  """When paginating backwards, are there more items?"""
  hasPreviousPage: Boolean!

  """When paginating backwards, the cursor to continue."""
  startCursor: String

  """When paginating forwards, the cursor to continue."""
  endCursor: String
}

"""The root of all... queries"""
type Query {
  """The ID of an object"""
  id: ID!

  """Fetches an object given its ID"""
  node(
    """The ID of an object"""
    id: ID
  ): Node

  """Fetches objects given their IDs"""
  nodes(
    """The IDs of objects"""
    ids: [ID!]!
  ): [Node]!
  event(id: ID!): Event
  events(after: String, first: Int, before: String, last: Int, search: String): EventConnection!
}
