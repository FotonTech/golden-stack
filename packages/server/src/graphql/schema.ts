import { GraphQLSchema } from 'graphql';

import MutationType from './type/MutationType';
import QueryType from './type/QueryType';

export const schema = new GraphQLSchema({
  query: QueryType,
  mutation: MutationType,
});
