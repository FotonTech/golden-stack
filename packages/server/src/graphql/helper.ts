import DataLoader from 'dataloader';

interface LoaderFactory<T extends DataLoader<any, any>> {
  getLoader: () => T;
}

interface LoaderFactoryMap {
  [key: string]: LoaderFactory<any>;
}

export type ResolvedLoaders<T extends LoaderFactoryMap> = { [K in keyof T]: ReturnType<T[K]['getLoader']> };

export function getDataloaders<T extends LoaderFactoryMap>(loaders: T): ResolvedLoaders<T> {
  const result: ResolvedLoaders<T> = {} as any;
  for (const key in loaders) {
    if (loaders[key].getLoader) {
      result[key] = loaders[key].getLoader();
    }
  }
  return result;
}
