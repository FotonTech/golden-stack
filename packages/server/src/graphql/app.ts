import { koaPlayground } from 'graphql-playground-middleware';
// import { print } from 'graphql/language';
import Koa, { Context } from 'koa';
import bodyParser from 'koa-bodyparser';
import convert from 'koa-convert';
import cors from 'koa-cors';
import graphqlHttp, { OptionsData } from 'koa-graphql';
import koaLogger from 'koa-logger';
import multer from 'koa-multer';
import Router from '@koa/router';

import * as loaders from '../loader';

import { GraphQLContext } from '../types';

import { getDataloaders } from './helper';
import { schema } from './schema';

const app = new Koa<any, Context>();
if (process.env.NODE_ENV === 'production') {
  app.proxy = true;
}

const router = new Router<any, Context>();

const storage = multer.memoryStorage();

const limits = {
  fieldSize: 30 * 1024 * 1024,
};

// if production than trick cookies library to think it is always on a secure request
if (process.env.NODE_ENV === 'production') {
  app.use((ctx, next) => {
    ctx.cookies.secure = true;
    return next();
  });
}

app.use(bodyParser());

app.use(async (ctx, next) => {
  try {
    await next();
  } catch (err) {
    // eslint-disable-next-line no-console
    console.log('koa error:', err);
    ctx.status = err.status || 500;
    ctx.app.emit('error', err, ctx);
  }
});

app.on('error', err => {
  // eslint-disable-next-line no-console
  console.error('Error while answering request', { error: err });
});

if (process.env.NODE_ENV !== 'test') {
  app.use(koaLogger());
}

app.use(convert(cors({ maxAge: 86400, origin: '*' })));

router.all('/graphql', multer({ storage, limits }).any());

router.all(
  '/playground',
  koaPlayground({
    endpoint: '/graphql',
  }),
);

// Middleware to get dataloaders
app.use((ctx, next) => {
  ctx.dataloaders = getDataloaders(loaders);
  return next();
});

router.all(
  '/graphql',
  convert(
    graphqlHttp(
      async (request, ctx, koaContext: unknown): Promise<OptionsData> => {
        const { dataloaders } = koaContext;
        const { appversion, appbuild, appplatform } = request.header;

        if (process.env.NODE_ENV !== 'test') {
          // eslint-disable-next-line no-console
          console.info('Handling request', {
            appversion,
            appbuild,
            appplatform,
          });
        }

        return {
          graphiql: process.env.NODE_ENV === 'development',
          schema,
          rootValue: {
            request: ctx.req,
          },
          context: {
            dataloaders,
            appplatform,
            koaContext,
          } as GraphQLContext,
          extensions: ({ document, variables, result }) => {
            // if (process.env.NODE_ENV === 'development') {
            //   if (document) {
            //     // eslint-disable-next-line no-console
            //     console.log(print(document));
            //     // eslint-disable-next-line no-console
            //     console.log(variables);
            //     // eslint-disable-next-line no-console
            //     console.log(JSON.stringify(result, null, 2));
            //   }
            // }
            return null as any;
          },
          formatError: (error: any) => {
            if (error.path || error.name !== 'GraphQLError') {
              // eslint-disable-next-line no-console
              console.error(error);
            } else {
              // eslint-disable-next-line no-console
              console.log(`GraphQLWrongQuery: ${error.message}`);
            }

            if (error.name && error.name === 'BadRequestError') {
              ctx.status = 400;
              ctx.body = 'Bad Request';
              return {
                message: 'Bad Request',
              };
            }

            // eslint-disable-next-line no-console
            console.error('GraphQL Error', { error });

            return {
              message: error.message,
              locations: error.locations,
              stack: error.stack,
            };
          },
        };
      },
    ),
  ),
);

app.use(router.routes()).use(router.allowedMethods());

export default app;
