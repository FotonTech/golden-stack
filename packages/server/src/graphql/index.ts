import 'core-js';
import { createServer } from 'http';

import { execute, subscribe } from 'graphql';
import { SubscriptionServer } from 'subscriptions-transport-ws';

import { GRAPHQL_PORT } from '../common/config';
import connectDatabase from '../common/database';

import app from './app';
import { schema } from './schema';

const runServer = async () => {
  try {
    // eslint-disable-next-line no-console
    console.log('connecting to database...');
    await connectDatabase();
  } catch (error) {
    // eslint-disable-next-line no-console
    console.error('Could not connect to database', { error });
    throw error;
  }

  const server = createServer(app.callback());

  server.listen(GRAPHQL_PORT, () => {
    // eslint-disable-next-line no-console
    console.info(`Server started on port: ${GRAPHQL_PORT}`);

    if (process.env.NODE_ENV !== 'production') {
      // eslint-disable-next-line no-console
      console.info(`GraphQL Playground available at /playground on port ${GRAPHQL_PORT}`);
    }

    SubscriptionServer.create(
      {
        // eslint-disable-next-line no-console
        onDisconnect: () => console.info('Client subscription disconnected'),
        execute,
        subscribe,
        schema,

        // eslint-disable-next-line no-console
        onConnect: connectionParams => console.info('Client subscription connected', connectionParams),
      },
      {
        server,
        path: '/subscriptions',
      },
    );
  });
};

(async () => {
  // eslint-disable-next-line no-console
  console.log('server starting...');
  await runServer();
})();
