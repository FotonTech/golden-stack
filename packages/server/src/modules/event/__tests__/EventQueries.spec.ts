import { graphql } from 'graphql';
import { toGlobalId } from 'graphql-relay';

import { schema } from '../../../graphql/schema';

import {
  clearDbAndRestartCounters,
  connectMongoose,
  createEvent,
  disconnectMongoose,
  getContext,
  sanitizeTestObject,
} from '../../../../test/helpers';

beforeAll(connectMongoose);

beforeEach(clearDbAndRestartCounters);

afterAll(disconnectMongoose);

describe('EventType queries', () => {
  it('should query an event', async () => {
    const event = await createEvent();

    // language=GraphQL
    const query = `
      query Q($id: ID) {
        event: node(id: $id) {
          id
          ... on Event {
            id
            title
            description
          }
        }
      }
    `;

    const variables = {
      id: toGlobalId('Event', event._id),
    };
    const rootValue = {};
    const context = await getContext();
    const result = await graphql(schema, query, rootValue, context, variables);

    expect(result.errors).toBeUndefined();
    expect(sanitizeTestObject(result)).toMatchSnapshot();
  });

  it('should query all events', async () => {
    await createEvent();
    await createEvent();
    await createEvent();
    await createEvent();
    await createEvent();

    // language=GraphQL
    const query = `
      query Q {
        events(first: 10) {
          edges {
            node {
              id
              title
              description
            }
          }
        }
      }
    `;

    const variables = {};
    const rootValue = {};
    const context = await getContext();
    const result = await graphql(schema, query, rootValue, context, variables);

    expect(result.errors).toBeUndefined();
    expect(sanitizeTestObject(result)).toMatchSnapshot();
  });

  it('should search all events', async () => {
    await createEvent({ description: 'desc one' });
    await createEvent({ title: 'title one' });
    await createEvent({ description: 'three' });
    await createEvent({ description: 'two' });
    await createEvent({ title: 'title two' });

    // language=GraphQL
    const query = `
      query Q {
        events(first: 10, search: "two") {
          edges {
            node {
              id
              title
              description
            }
          }
        }
      }
    `;

    const variables = {};
    const rootValue = {};
    const context = await getContext();
    const result = await graphql(schema, query, rootValue, context, variables);

    expect(result.errors).toBeUndefined();
    expect(sanitizeTestObject(result)).toMatchSnapshot();
  });
});
