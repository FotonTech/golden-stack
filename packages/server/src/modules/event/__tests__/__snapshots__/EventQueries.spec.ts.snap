// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`EventType queries should query all events 1`] = `
Object {
  "data": Object {
    "events": Object {
      "edges": Array [
        Object {
          "node": Object {
            "description": "This is an awesome event #1",
            "id": "FROZEN-ID",
            "title": "Event #1",
          },
        },
        Object {
          "node": Object {
            "description": "This is an awesome event #2",
            "id": "FROZEN-ID",
            "title": "Event #2",
          },
        },
        Object {
          "node": Object {
            "description": "This is an awesome event #3",
            "id": "FROZEN-ID",
            "title": "Event #3",
          },
        },
        Object {
          "node": Object {
            "description": "This is an awesome event #4",
            "id": "FROZEN-ID",
            "title": "Event #4",
          },
        },
        Object {
          "node": Object {
            "description": "This is an awesome event #5",
            "id": "FROZEN-ID",
            "title": "Event #5",
          },
        },
      ],
    },
  },
}
`;

exports[`EventType queries should query an event 1`] = `
Object {
  "data": Object {
    "event": Object {
      "description": "This is an awesome event #1",
      "id": "FROZEN-ID",
      "title": "Event #1",
    },
  },
}
`;

exports[`EventType queries should search all events 1`] = `
Object {
  "data": Object {
    "events": Object {
      "edges": Array [
        Object {
          "node": Object {
            "description": "two",
            "id": "FROZEN-ID",
            "title": "Event #4",
          },
        },
        Object {
          "node": Object {
            "description": "This is an awesome event #5",
            "id": "FROZEN-ID",
            "title": "title two",
          },
        },
      ],
    },
  },
}
`;
